### DISCUSSION TITLE
Change pair and add new pair

### DISCUSSION THREAD

**Discussion by k4leg** [Fuco1/smartparens#1126] [Q&A] [answered]
Hi, I trying to change some predefined LaTeX quotes: `<<` and `>>` instead of ``​`​`​`` and `''`, and define new pair ``"`​`` and `"'` with tiggers `"` and `'` resp.  I tried this:
```emacs-lisp
(require 'smartparens-config)
(smartparens-global-mode)
(sp-with-modes '(tex-mode plain-tex-mode latex-mode LaTeX-mode)
  (sp-local-pair "`" "'" :actions nil)  ; Remove this pair.
  (sp-local-pair "``" "''" :actions nil)  ; Remove this pair.
  ; New pair that appears when I press apostrophe char.
  (sp-local-pair "\"`" "\"'"
    :trigger "'"
    :unless '(sp-latex-point-after-backslash sp-in-math-p)
    :post-handlers '(sp-latex-skip-double-quote))
  ; New pair that appears when I press double quote char.
  (sp-local-pair "<<" ">>"
    :trigger "\""
    :unless '(sp-latex-point-after-backslash sp-in-math-p)
    :post-handlers '(sp-latex-skip-double-quote))
  (sp-local-tag "\"" "<<" ">>" :actions '(wrap))
  (sp-local-tag "'" "\"`" "\"'" :actions '(wrap)))
```

But this code doesn't work (`|` is cursor):
- when I press `"` I see ``​`​`|`` instead of `<<|>>`
- when I press `"` twice (i. e. `""`) I see `<<|>>`
- `'` behavior is terrible: when I press `'` in
  ```latex
  \begin{document}
  |
  \end{document}
  ```
  I see
  ```latex
  \begin{document"'
  `|
  \end{document}
  ```

What I want:
- when I press `"` I see `<<|>>`
- when I press `'` I see ``"`|"'``
- when I press `'` in `<<Some text: |>>` I see ``<<Some text: "`|"'\,>>`` (`\,` is added between `"'` and `>>`)
- when I press `'` in `<<| is \ldots>>` I see ``<<\,"`|"' is \ldots>>`` (`\,` is added between `<<` and ``"`​``)
- vice versa `\,` doesn't work, i. e. there's no ``"`\,<<Some text>>, \ldots"'`` or ``"`\ldots <<Some text>>\,"'`` situation: `<<` and `>>` is outer, ``"`​`` and `"'` is inner quotes

At least I want first and second (without `\,`).

Additional info:
- [Full config](http://0x0.st/omo1.el)
- `smartparens` version: 20220204.1134
- Active `major-mode`: `latex-mode`
- Smartparens strict mode: nil
- Emacs version (`M-x emacs-version`): GNU Emacs 28.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.17.6) of 2022-04-28
- Starterkit/Distribution: Evil
- OS: Arch Linux

**Comment by Fuco1** [comment, marked as answer]
Hi.  Unfortunatelly the `sp-tags` functionality was removed couple years ago and there is no replacement.

For the `"` to be used as a trigger you must unbind `TeX-insert-quote` because it does its own processing.

Then you must disable `"` and `'` pairs because they have special handling of string syntax and it messes things up.  This is a latex compat issue.

```elisp
(sp-local-pair 'latex-mode "'" "'" :actions nil)
(sp-local-pair 'latex-mode "\"" "\"" :actions nil)
```

After this, it works fine.  For the `\,` you can setup a post handler.
